gui/wxpython/gui_core/menu.py

```diff
@@ -305,3 +307,121 @@ def OnItemSelected(self, node):
+        # Recent files path stored in GRASS GIS config dir in the
+        # .recent_files file in the group by application name
+        self.config = wx.FileConfig(
+            style=wx.CONFIG_USE_LOCAL_FILE,
+            localFilename=os.path.join(
+                utils.GetSettingsPath(), self.recent_files,
+            ),
+        )
+        self.config.SetPath(strPath=app_name)
+        self.filehistory.Load(self.config)
+        self.RemoveFileFromHistory(stop=self.filehistory.GetCount())
+
+        self.recent = wx.Menu()
+        self.filehistory.UseMenu(self.recent)
+        self.filehistory.AddFilesToMenu()
+
+        # Show recent files menu if count of items in menu > 0
+        if self.filehistory.GetCount() > 0:
+            self._insertMenu()
+
+    def _insertMenu(self):
+        """Insert recent files menu into the parent menu on the
+        specified position if count of menu items > 0"""
+
+        self._parent_menu.Insert(
+            pos=self._pos, id=wx.ID_ANY, text=_('&Recent Files'),
+            submenu=self.recent,
+        )
+        self.recent.Bind(
+            wx.EVT_MENU_RANGE, self._onFileHistory,
+            id=wx.ID_FILE1, id2=wx.ID_FILE + self._history_len,
+        )
+
+    def _onFileHistory(self, event):
+        """Choose recent file from menu event"""
+
+        file_exists = True
+        file_index = event.GetId() - wx.ID_FILE1
+        path = self.filehistory.GetHistoryFile(file_index)
+
+        if not os.path.exists(path):
+            self.RemoveFileFromHistory(
+                start=file_index, stop=file_index + 1,
+            )
+            file_exists = False
+        self.file_requested.emit(
+            path=path, file_exists=file_exists,
+            file_history=self.filehistory,
+        )
+
+    def AddFileToHistory(self, filename):
+        """Add file to history, and save history into '.recent_files'
+        file
+
+        :param str filename: file path
+
+        :return None
+        """
+
+        if self.filehistory.GetCount() == 0:
+            self._insertMenu()
+        if filename:
+            self.filehistory.AddFileToHistory(filename)
+            self.filehistory.Save(self.config)
+            self.config.Flush()
+
+    def RemoveFileFromHistory(self, start=0, stop=1):
+        """Remove file from history and from '.recent_files' file
+
+        :param int start: start loop index
+        :param int stop: stop loop index
+
+        :return: None
+        """
+
+        file_index = start
+        while (file_index < stop):
+            file = self.filehistory.GetHistoryFile(index=file_index)
+            if not os.path.exists(file):
+                self.filehistory.RemoveFileFromHistory(i=file_index)
+                stop = self.filehistory.GetCount()
+                file_index = start
+            else:
+                file_index += 1
+        self.filehistory.Save(self.config)
+        self.config.Flush()
```

Comment:
Is this function meant to remove non-existent files or files specified by range? It seems to be used to be used to remove file by index, but it removes only non-existent files and when it does it stretches its search parameters.

Reply:
> Is this function meant to remove non-existent files or files specified by range? It seems to be used to be used to remove file by index, but it removes only non-existent files and when it does it stretches its search parameters.

In the [84ac667](https://github.com/OSGeo/grass/pull/1021/commits/84ac667328fde707251329beeb1d564e465f50ab) method is improved.

e.g.

- remove one file (file index in the history file is 2): start=2, end=2 + 1
- remove all files (e.g. max files count in the history is 2): end=2

added `check_if_files_exists=False` parameter (allow remove file(s) only, which path doesn't exists)